node/src/components/storage.rs

```diff
@@ -1134,57 +1154,59 @@ impl Storage {
             .map_err(Into::into)
     }
 
-    /// Writes a single block body in a separate transaction to storage.
-    fn put_single_block_body_v2(
+    fn put_merkle_block_body_part<'a, T>(
         &self,
         tx: &mut RwTransaction,
-        block_body_hash: &Digest,
-        block_body: &BlockBody,
-    ) -> Result<bool, LmdbExtError> {
-        let hashable_parts = block_body.hashed_parts();
-        let BlockBodyHashedParts {
-            hashed_proposer,
-            hashed_deploy_hashes,
-            hashed_transfer_hashes,
-        } = &hashable_parts;
-
-        // Insert parts and the vec of hashes to new dbs
+        part_database: Database,
+        merklized_block_body_part: &MerkleBlockBodyPart<'a, T>,
+    ) -> Result<bool, LmdbExtError>
+    where
+        T: Serialize,
+    {
+        // It's possible the value is already present (ie, if it is a block proposer).
+        // We put the value and rest hashes in first, since we need that present even if the value
+        // is already there.
         if !tx.put_value(
-            self.proposer_db,
-            hashed_proposer,
-            block_body.proposer(),
+            self.block_body_v2_db,
+            merklized_block_body_part.merkle_linked_list_node_hash(),
+            &merklized_block_body_part.value_and_rest_hashes_slice(),
             true,
         )? {
             return Ok(false);
-        }
+        };
 
         if !tx.put_value(
-            self.deploy_hashes_db,
-            hashed_deploy_hashes,
-            block_body.deploy_hashes(),
+            part_database,
+            merklized_block_body_part.value_hash(),
+            merklized_block_body_part.value(),
             true,
         )? {
             return Ok(false);
-        }
+        };
+        Ok(true)
+    }
 
-        if !tx.put_value(
-            self.transfer_hashes_db,
-            hashed_transfer_hashes,
-            block_body.transfer_hashes(),
-            true,
-        )? {
+    /// Writes a single merklized block body in a separate transaction to storage.
+    fn put_single_block_body_v2(
+        &self,
+        tx: &mut RwTransaction,
+        block_body: &BlockBody,
+    ) -> Result<bool, LmdbExtError> {
+        let merkle_block_body = block_body.merklize();
+        let MerkleBlockBody {
+            deploy_hashes,
+            transfer_hashes,
+            proposer,
+        } = &merkle_block_body;
+        if !self.put_merkle_block_body_part(tx, self.deploy_hashes_db, deploy_hashes)? {
             return Ok(false);
-        }
-
-        if !tx.put_value(
-            self.block_body_v2_db,
-            &block_body_hash,
-            &hashable_parts.as_slice(),
-            true,
-        )? {
+        };
+        if !self.put_merkle_block_body_part(tx, self.transfer_hashes_db, transfer_hashes)? {
             return Ok(false);
-        }
-
+        };
+        // This might already be present, don't bother checking and early returning.
```

Comment:
Does `Ok(false)` mean that the value was present? If yes, then other early returns might be incorrect. Empty blocks with the same proposer will have all of the hashes exactly equal (the Merkle tree leaves will be empty list hash, empty list hash and proposer hash - so the same Merkle root if proposer is the same). This means we shouldn't treat it as a problem if we have a collision in deploys or transfers, either.

Reply:
As discussed, since we are overwriting these should always come back `true`.

I've reverted to the old behavior in https://github.com/casper-network/casper-node/pull/1563/commits/f4e5eb89a47ce6e6334bceab556a4297fc943316